query/query/query.py

```diff
@@ -52,107 +66,97 @@ async def healthcheck(request):  # pylint: disable=unused-argument
-    return java_to_web_response(jresp)
+def blocking_value_type(jbackend, userdata, body):
+    return jbackend.valueType(userdata['username'], body['code'])
 
 
-def blocking_value_type(jbackend, username, code):
-    return jbackend.valueType(username, code)
+def blocking_table_type(jbackend, userdata, body):
+    return jbackend.tableType(userdata['username'], body['code'])
 
 
-@routes.post('/type/value')
-@rest_authenticated_users_only
-async def value_type(request, userdata):
-    app = request.app
-    thread_pool = app['thread_pool']
-    jbackend = app['jbackend']
-    code = await request.json()
-    log.info(f'value type: {code}')
-    await add_user(app, userdata)
-    jresp = await blocking_to_async(thread_pool, blocking_value_type, jbackend, userdata['username'], code)
-    return java_to_web_response(jresp)
+def blocking_matrix_type(jbackend, userdata, body):
+    return jbackend.matrixTableType(userdata['username'], body['code'])
 
 
-def blocking_table_type(jbackend, username, code):
-    return jbackend.tableType(username, code)
+def blocking_blockmatrix_type(jbackend, userdata, body):
+    return jbackend.blockMatrixType(userdata['username'], body['code'])
 
 
-@routes.post('/type/table')
-@rest_authenticated_users_only
-async def table_type(request, userdata):
+def blocking_get_reference(jbackend, userdata, body):   # pylint: disable=unused-argument
+    return jbackend.referenceGenome(userdata['username'], body['name'])
+
+
+async def handle_ws_response(request, userdata, cmd_str, f):
+    log.info('connecting websocket')
     app = request.app
-    thread_pool = app['thread_pool']
     jbackend = app['jbackend']
-    code = await request.json()
-    log.info(f'table type: {code}')
+
     await add_user(app, userdata)
-    jresp = await blocking_to_async(thread_pool, blocking_table_type, jbackend, userdata['username'], code)
-    return java_to_web_response(jresp)
+    log.info('connecting websocket')
+    ws = web.WebSocketResponse(heartbeat=30, max_msg_size=0)
+    task = None
+    try:
+        await ws.prepare(request)
+        app['sockets'].add(ws)
+        log.info(f'websocket prepared: {ws}')
+        body = await ws.receive_json()
+
+        log.info(f"{cmd_str}: {body}")
+        await ws.send_str(cmd_str)
+        task = asyncio.ensure_future(send_ws_response(app['thread_pool'], ws, f, jbackend, userdata, body))
+        r = await ws.receive()
+        log.info('Received websocket message. Expected CLOSE, got {r}')
+        assert r.type == aiohttp.WSMsgType.CLOSE
+        return ws
+    finally:
+        if not ws.closed:
+            await ws.close()
+            log.info('Websocket was not closed. Closing.')
+        if task is not None and not task.done():
+            task.cancel()
+            log.info('Task has been cancelled due to websocket closure.')
+        log.info('websocket connection closed')
+        app['sockets'].remove(ws)
```

Comment:
I don't think we need this. handle_ws_response is only invoked by route handlers. Route handlers are cancelled when the server thread receives an interrupt. We can't leak web sockets because of the try-finally.